docs/proposals/formatted-output.md

```diff
@@ -0,0 +1,359 @@
+```
+
+### oras push
+
+When using `oras push` with the flag `--format`, the following fields should be formatted into JSON output:
+
+- `reference`: full artifact reference by digest, e.g, `$REGISTRY/$REPO@$DIGEST`
+- `referenceByTags`: array, pushed tags by reference, e.g. `$REGISTRY/$REPO@TAG1`
+- `mediaType`: media type of the pushed file (layer) 
+- `digest`: digest of the pushed file (layer) 
+- `size`: file size in bytes
+- `annotations`: contains arbitrary metadata for the image manifest
+
+For example, push a file and two tags to a repository and show the descriptor of the image manifest in pretty JSON format. 
+
+```bash
+oras push $REGISTRY/$REPO:$TAG1, $TAG2 sbom.spdx vul-scan.json --format json 
+```
+
+```json
+{
+  "reference": "localhost:5000/oras@sha256:4a5b8c83d153f52afdfcb422db56c2349aae3bd5ecf8338a58353b5eb6681c45",
+  "mediaType": "application/vnd.oci.image.manifest.v1+json",
+  "digest": "sha256:4a5b8c83d153f52afdfcb422db56c2349aae3bd5ecf8338a58353b5eb6681c45",
+  "size": 820,
+  "annotations": {
+    "org.opencontainers.image.created": "2023-12-15T09:41:54Z"
+  },
+  "artifactType": "json/example",
+  "referenceByTags": [
+    "$REGISTRY/$REPO:$TAG1",
+    "$REGISTRY/$REPO:$TAG2"
+  ]
+}
+```
+
+> [!NOTE]
+> When pushing a folder to filesystem, the output fields are the same as the example of pushing files as above.
+
+Push a folder to a repository and filter out the value of `reference` and `artifactType` of the pushed artifact in the standard output.
+
+```bash
+oras push $REGISTRY/$REPO:$TAG sample-folder --format '{{.reference}}, {{.mediaType}}'
+```
+
+```console
+localhost:5000@sha256:85438e6598bf35057962fff34399a362d469ca30a317939427fca6b7a289e70d, application/vnd.oci.image.manifest.v1+json
+```
+
+### oras attach
+
+When using `oras pull` with the flag `--format`, the following fields should be formatted into JSON output:
+
+- `reference`: full reference by digest of the referrer file
+- `mediaType`: media type of the referrer
+- `size`: referrer file size in bytes
+- `digest`: digest of the attached referrer file
+- `artifactType`: artifact type of the referrer
+- `annotations`: contains arbitrary metadata in a referrer
+
+For example, attach two files to an image and show the descriptor metadata of the referrer in JSON format.
+
+```bash
+oras attach $REGISTRY/$REPO:$TAG --artifact-type example/report-and-sbom vul-report.json:example/vul-scan sbom.spdx:example/sbom --format json
+```
+
+```json
+{
+  "tags": "$REGISTRY/$REPO:$TAG",
```

Comment:
This example should be an array.

Reply:
Good catch. The `referenceByTags` only exists in `oras push`. I removed this property from `oras attach`.